cpp/src/rolling/rolling_detail.cuh

```diff
@@ -1062,6 +1062,72 @@ struct rolling_window_launcher {
     return gather_map;
   }
 
+  /**
+   * @brief Purge entries for null inputs from gather_map, and adjust offsets.
+   */
+  void purge_null_entries(column_view const& input,
+                          std::unique_ptr<column>& gather_map,
+                          std::unique_ptr<column>& offsets,
+                          rmm::cuda_stream_view stream,
+                          rmm::mr::device_memory_resource* mr)
+  {
+    auto input_device_view = column_device_view::create(input, stream);
+
+    auto input_row_is_null = [d_input = *input_device_view] __device__(auto i) {
+      return d_input.is_null_nocheck(i);
+    };
+    auto input_row_not_null = [d_input = *input_device_view] __device__(auto i) {
+      return d_input.is_valid_nocheck(i);
+    };
+
+    auto num_child_nulls = thrust::count_if(rmm::exec_policy(stream),
+                                            gather_map->view().template begin<size_type>(),
+                                            gather_map->view().template end<size_type>(),
+                                            input_row_is_null);
+
+    if (num_child_nulls == 0) {
+      return;  // No child nulls. Nothing to recompute.
+    }
+
+    // Purge entries in gather_map that correspond to null input.
+    auto new_gather_map = make_fixed_width_column(data_type{type_to_id<size_type>()},
+                                                  gather_map->size() - num_child_nulls,
+                                                  mask_state::UNALLOCATED,
+                                                  stream,
+                                                  mr);
+    thrust::copy_if(rmm::exec_policy(stream),
+                    gather_map->view().template begin<size_type>(),
+                    gather_map->view().template end<size_type>(),
+                    new_gather_map->mutable_view().template begin<size_type>(),
+                    input_row_not_null);
+
+    // Recalculate offsets after null entries are purged.
+    auto new_sizes = make_fixed_width_column(
+      data_type{type_to_id<size_type>()}, input.size(), mask_state::UNALLOCATED, stream, mr);
+
+    thrust::transform(rmm::exec_policy(stream),
+                      thrust::make_counting_iterator<size_type>(0),
+                      thrust::make_counting_iterator<size_type>(input.size()),
+                      new_sizes->mutable_view().template begin<size_type>(),
+                      [d_gather_map  = gather_map->view().template begin<size_type>(),
+                       d_old_offsets = offsets->view().template begin<size_type>(),
+                       input_row_not_null] __device__(auto i) {
+                        return thrust::count_if(thrust::seq,
+                                                d_gather_map + d_old_offsets[i],
+                                                d_gather_map + d_old_offsets[i + 1],
+                                                input_row_not_null);
+                      });
```

Comment:
This looks like it could likewise be replaced by `segmented_count_unset_bits` (with some modifications) https://docs.rapids.ai/api/libcudf/nightly/group__column__nullmask.html#ga81838307b321a8a911054944a74f4bfb

Reply:
Thanks for the tip. I'll try rephrase this to use `segmented_count_set_bits()`.

Reply:
`segmented_count_set_bits()` takes a `std::vector` of segment offset (pairs). This implies that the vector would have nearly twice as many elements as there are list rows in the column. 
Pardon my naiveté, but is it permitted to copy the offsets column over to host memory (and doubled) to use `segmented_count_unset_bits()`?

Reply:
Actually, no, nix that.
For the same reason as [described above](https://github.com/rapidsai/cudf/pull/7264#discussion_r568112965), it doesn't look like the offset recomputation can be expressed as a call to `segmented_count_set_bits()`, without it being convoluted. :/
In short, we're not counting valid rows in the `gather_map`; we're counting rows in the `gather_map` that point to valid rows in the `input` column.